File: restAPI/helper_functions/kafka-helper-functions.js
const { Kafka } = require("kafkajs");
// KAFKA TEST
// Kafka broker connection details
const kafkaBrokers = ["localhost:9092"];
const kafkaTopic = "test";

// Create a new Kafka producer
const kafka = new Kafka({
  clientId: "quickstart--shared-admin",
  brokers: kafkaBrokers,
});
const producer = kafka.producer({ allowAutoTopicCreation: true });

/**
 * Function to launch Kafka Connect and establish connection with Neo4j
 */
function setupKafkaConnect() {
  fetch("http://localhost:8083/connectors", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
    },
    body: JSON.stringify({
      name: "Neo4jSinkConnectorJSONString",
      config: {
        topics: "test",
        "connector.class": "streams.kafka.connect.sink.Neo4jSinkConnector",
        "key.converter": "org.apache.kafka.connect.json.JsonConverter",
        "key.converter.schemas.enable": false,
        "value.converter": "org.apache.kafka.connect.json.JsonConverter",
        "value.converter.schemas.enable": false,
        "errors.retry.timeout": "-1",
        "errors.retry.delay.max.ms": "1000",
        "errors.tolerance": "all",
        "errors.log.enable": true,
        "errors.log.include.messages": true,
        "neo4j.server.uri": "bolt://neo4jServer:7687",
        "neo4j.authentication.basic.username": "neo4j",
        "neo4j.authentication.basic.password": "password",
        "neo4j.topic.cud": "test",
      },
    }),
  })
    .then((res) => {
      console.log("Connector set!");
    })
    .catch((err) => {
      console.log("Error when setting connector: " + err);
    });
}

/**
 * Function to create Dag node on Neo4j
 * @param {String} dagId - The ID of the Dag to be created
 */
async function createDagNode(dagId) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATEDAG MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {
                dagId,
              },
              ids: { dagId },
              labels: ["Dag"],
              type: "node",
              detach: true,
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log("Error when creating: " + dagId + " Error message: " + err);
    });
}

/**
 * Function to persist "is_parent_of" relationship between Dag and Task on Neo4j
 * @param {String} dagId - The ID of the parent Dag
 * @param {String} taskId - The ID of the task belonging to the parent Dag
 */
async function createDagTaskRelationship(dagId, taskId) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATEDAGTASKRS MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {},
              rel_type: "is_parent_of",
              from: {
                ids: { dagId },
                labels: ["Dag"],
                op: "merge",
              },
              to: {
                ids: { taskId: dagId + "." + taskId },
                labels: ["Job"],
                op: "merge",
              },
              type: "relationship",
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " +
          dagId +
          " to " +
          taskId +
          " Error message: " +
          err
      );
    });
}

/**
 * Function to persist "activates" relationship between 2 Tasks on Neo4j to illustrate dependency
 * @param {String} taskId1 - The ID of the upstream Task
 * @param {String} taskId2 - The ID of the downstream Task
 */
async function createTaskTaskRelationship(taskId1, taskId2) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATETASKTASKRS MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {},
              rel_type: "activates",
              from: {
                ids: { taskId: taskId1 },
                labels: ["Job"],
                op: "merge",
              },
              to: {
                ids: { taskId: taskId2 },
                labels: ["Job"],
                op: "merge",
              },
              type: "relationship",
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " +
          taskId1 +
          " to " +
          taskId2 +
          " Error message: " +
          err
      );
    });
}

async function createSparkJobNode(sparkJobId) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATESPARKJOB MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {
                sparkJobId,
              },
              ids: { sparkJobId },
              labels: ["Spark Job"],
              type: "node",
              detach: true,
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " + sparkJobId + " Error message: " + err
      );
    });
}

async function createTaskSparkJobRelationship(taskId, sparkJobId) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATETASKSPARKJOBRS MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {},
              rel_type: "activates",
              from: {
                ids: { taskId },
                labels: ["Job"],
                op: "merge",
              },
              to: {
                ids: { sparkJobId },
                labels: ["Spark Job"],
                op: "merge",
              },
              type: "relationship",
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " +
          taskId +
          " to " +
          sparkJobId +
          " Error message: " +
          err
      );
    });
}

async function createSparkJobSparkTaskRelationship(
  parentSparkJobId,
  sparkTaskId
) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATESPARKJOBSPARKTASKRS MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {},
              rel_type: "parent_of",
              from: {
                ids: { sparkJobId: parentSparkJobId },
                labels: ["Spark Job"],
                op: "merge",
              },
              to: {
                ids: { sparkTaskId: sparkTaskId },
                labels: ["Spark Task"],
                op: "merge",
              },
              type: "relationship",
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " +
          parentSparkJobId +
          " to " +
          sparkTaskId +
          " Error message: " +
          err
      );
    });
}

async function createDatasetToSparkTaskRelationship(datasetId, sparkTaskId) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATEDATASETTOSPARKTASKRS MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {},
              rel_type: "used_in",
              from: {
                ids: { datasetId },
                labels: ["Dataset"],
                op: "merge",
              },
              to: {
                ids: { sparkTaskId },
                labels: ["Spark Task"],
                op: "merge",
              },
              type: "relationship",
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " +
          datasetId +
          " to " +
          sparkTaskId +
          " Error message: " +
          err
      );
    });
}

async function createSparkTaskToDatasetRelationship(sparkTaskId, datasetId) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATESPARKTASKDATASETRS MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {},
              rel_type: "outputs_to",
              from: {
                ids: { sparkTaskId },
                labels: ["Spark Task"],
                op: "merge",
              },
              to: {
                ids: { datasetId },
                labels: ["Dataset"],
                op: "merge",
              },
              type: "relationship",
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " +
          sparkTaskId +
          " to " +
          datasetId +
          " Error message: " +
          err
      );
    });
}

async function createDatasetToTaskRelationship(datasetId, taskId) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATEDATASETTOTASKRS MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {},
              rel_type: "used_in",
              from: {
                ids: { datasetId },
                labels: ["Dataset"],
                op: "merge",
              },
              to: {
                ids: { taskId },
                labels: ["Task"],
                op: "merge",
              },
              type: "relationship",
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " +
          datasetId +
          " to " +
          taskId +
          " Error message: " +
          err
      );
    });
}

async function createTaskToDatasetRelationship(taskId, datasetId) {
  producer
    .connect()
    .then(() => {
      console.log("SENDING CREATETASKDATASETRS MESSAGE TO Q");
      producer.send({
        topic: "test",
        messages: [
          {
            value: JSON.stringify({
              op: "merge",
              properties: {},
              rel_type: "outputs_to",
              from: {
                ids: { taskId },
                labels: ["Task"],
                op: "merge",
              },
              to: {
                ids: { datasetId },
                labels: ["Dataset"],
                op: "merge",
              },
              type: "relationship",
            }),
          },
        ],
      });
    })
    .catch((err) => {
      console.log(
        "Error when creating: " +
          taskId +
          " to " +
          datasetId +
          " Error message: " +
          err
      );
    });
}

module.exports = {
  setupKafkaConnect,
  createDagNode,
  createDagTaskRelationship,
  createTaskTaskRelationship,
  createSparkJobNode,
  createSparkJobSparkTaskRelationship,
  createTaskSparkJobRelationship,
  createDatasetToSparkTaskRelationship,
  createSparkTaskToDatasetRelationship,
  createDatasetToTaskRelationship,
  createTaskToDatasetRelationship,
};

File: Python Lineage Creation/helper_functions/kafka_helper_functions.py
from kafka import KafkaProducer
import requests
import json
import asyncio

kafka_connect_url = "http://localhost:8083/connectors"
kafka_server = "localhost:9092"
kafka_topic = "test"

producer = KafkaProducer(bootstrap_servers=[kafka_server], client_id="quickstart--shared-admin", value_serializer= lambda v: bytes(json.dumps(v).encode('utf-8')))
def setup_kafka_connect():
    headers = {
        "Content-Type": "application/json",
        "Accept": "application/json"
    }
    payload = {
        "name": "Neo4jSinkConnectorJSONString",
        "config": {
            "topics": "test",
            "connector.class": "streams.kafka.connect.sink.Neo4jSinkConnector",
            "key.converter": "org.apache.kafka.connect.json.JsonConverter",
            "key.converter.schemas.enable": False,
            "value.converter": "org.apache.kafka.connect.json.JsonConverter",
            "value.converter.schemas.enable": False,
            "errors.retry.timeout": "-1",
            "errors.retry.delay.max.ms": "1000",
            "errors.tolerance": "all",
            "errors.log.enable": True,
            "errors.log.include.messages": True,
            "neo4j.server.uri": "bolt://neo4jServer:7687",
            "neo4j.authentication.basic.username": "neo4j",
            "neo4j.authentication.basic.password": "password",
            "neo4j.topic.cud": "test"
        }
    }

    try:
        response = requests.post(kafka_connect_url, headers=headers, data=json.dumps(payload))
        print("Connector set!")
    except requests.exceptions.RequestException as err:
        print("Error when setting connector: " + str(err))

async def create_dag_node(dagId):
    try:
        print("SENDING CREATEDAG MESSAGE TO Q")
        producer.send(
            kafka_topic,
            value = {
                "op": "merge",
                "properties": {
                    "dagId": dagId,
                },
                "ids": {"dagId": dagId},
                "labels": ["Dag"],
                "type": "node",
                "detach": True,
                }
        )
        producer.flush()
    except Exception as err:
        print(f"Error when creating: {dagId} Error message: {err}")

async def create_dag_task_relationship(dag_id, task_id):
    try:
        print("SENDING CREATEDAGTASKRS MESSAGE TO Q")
        data = {
                "op": "merge",
                "properties": {},
                "rel_type": "is_parent_of",
                "from": {
                    "ids": {"dagId": dag_id},
                    "labels": ["Dag"],
                    "op": "merge"
                },
                "to": {
                    "ids": {"taskId": f"{dag_id}.{task_id}"},
                    "labels": ["Airflow Task"],
                    "op": "merge"
                },
                "type": "relationship"
                }
        producer.send(kafka_topic, value=data)
        producer.flush()
    except Exception as err:
        print(
        "Error when creating: " +
          dag_id +
          " to " +
          task_id +
          " Error message: " +
          err
      )

async def create_dag_dag_relationship(dag_id_1, dag_id_2):
    try:
        print("SENDING CREATEDAGDAG MESSAGE TO Q")
        data = {
              "op": "merge",
              "properties": {},
              "rel_type": "triggers",
              "from": {
                "ids": { "dagId": dag_id_1 },
                "labels": ["Dag"],
                "op": "merge",
              },
              "to": {
                "ids": { "dagId": dag_id_2 },
                "labels": ["Dag"],
                "op": "merge",
              },
              "type": "relationship",
            } 
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          dag_id_1 +
          " to " +
          dag_id_2 +
          " Error message: " +
          err)

async def create_task_task_relationship(task_id_1, task_id_2):
    try:
        print("SENDING CREATEDAGDAG MESSAGE TO Q")
        data = {
              "op": "merge",
              "properties": {},
              "rel_type": "activates",
              "from": {
                "ids": { "taskId": task_id_1 },
                "labels": ["Airflow Task"],
                "op": "merge",
              },
              "to": {
                "ids": { "taskId": task_id_2 },
                "labels": ["Airflow Task"],
                "op": "merge",
              },
              "type": "relationship",
            } 
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          task_id_1 +
          " to " +
          task_id_2 +
          " Error message: " +
          err)

async def create_spark_job_node(spark_job_id):
    try:
        print("SENDING CREATESPARKJOB MESSAGE TO Q")
        data ={
              "op": "merge",
              "properties": {
                "sparkJobId" : spark_job_id,
              },
              "ids": { "sparkJobId": spark_job_id },
              "labels": ["Spark Job"],
              "type": "node",
              "detach": True,
            }
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          spark_job_id +
          " Error message: " +
          err)

async def create_task_spark_job_relationship(task_id, spark_job_id):
    try:
        print("SENDING CREATETASKSPARKJOBRS MESSAGE TO Q")
        data = {
              "op": "merge",
              "properties": {},
              "rel_type": "activates",
              "from": {
                "ids": { "taskId": task_id },
                "labels": ["Airflow Task"],
                "op": "merge",
              },
              "to": {
                "ids": { "sparkJobId": spark_job_id },
                "labels": ["Spark Job"],
                "op": "merge",
              },
              "type": "relationship",
            } 
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          task_id +
          " to " +
          spark_job_id +
          " Error message: " +
          err)

async def create_spark_job_spark_task_relationship(parent_spark_job_id, spark_task_id):
    try:
        print("SENDING CREATESPARKJOBSPARKTASKRS MESSAGE TO Q")
        data ={
              "op": "merge",
              "properties": {},
              "rel_type": "parent_of",
              "from": {
                "ids": { "sparkJobId": parent_spark_job_id },
                "labels": ["Spark Job"],
                "op": "merge",
              },
              "to": {
                "ids": { "sparkTaskId": spark_task_id },
                "labels": ["Spark Task"],
                "op": "merge",
              },
              "type": "relationship",
            } 
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          parent_spark_job_id +
          " to " +
          spark_task_id + 
          " Error message: " +
          err)

async def create_dataset_spark_task_relationship(dataset_id, spark_task_id):
    try:
        print("SENDING CREATEDATASETTOSPARKTASKRS MESSAGE TO Q")
        data ={
              "op": "merge",
              "properties": {},
              "rel_type": "used_in",
              "from": {
                "ids": { "datasetId": dataset_id },
                "labels": ["Dataset"],
                "op": "merge",
              },
              "to": {
                "ids": { "sparkTaskId": spark_task_id },
                "labels": ["Spark Task"],
                "op": "merge",
              },
              "type": "relationship",
            } 
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          dataset_id +
          " to " +
          spark_task_id + 
          " Error message: " +
          err)

async def create_spark_task_dataset_relationship(spark_task_id, dataset_id):
    try:
        print("SENDING CREATESPARKTASKDATASETRS MESSAGE TO Q")
        data ={
              "op": "merge",
              "properties": {},
              "rel_type": "outputs_to",
               "from": {
                "ids": { "sparkTaskId": spark_task_id },
                "labels": ["Spark Task"],
                "op": "merge",
              },
              "to": {
                "ids": { "datasetId": dataset_id },
                "labels": ["Dataset"],
                "op": "merge",
              },
              "type": "relationship",
            } 
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          spark_task_id +
          " to " +
          dataset_id + 
          " Error message: " +
          err)

async def create_dataset_task_relationship(dataset_id, task_id):
    try:
        print("SENDING CREATEDATASETTOTASKRS MESSAGE TO Q")
        data ={
              "op": "merge",
              "properties": {},
              "rel_type": "used_in",
               "from": {
                "ids": { "datasetId": dataset_id },
                "labels": ["Dataset"],
                "op": "merge",
              },
              "to": {
                "ids": { "taskId": task_id },
                "labels": ["Airflow Task"],
                "op": "merge",
              },
              "type": "relationship",
            } 
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          dataset_id +
          " to " +
          task_id + 
          " Error message: " +
          err)

async def create_task_dataset_relationship(task_id, dataset_id):
    try:
        print("SENDING CREATESPARKTASKDATASETRS MESSAGE TO Q")
        data ={
              "op": "merge",
              "properties": {},
              "rel_type": "outputs_to",
               "from": {
                "ids": { "taskId": task_id },
                "labels": ["Airflow Task"],
                "op": "merge",
              },
              "to": {
                "ids": { "datasetId": dataset_id },
                "labels": ["Dataset"],
                "op": "merge",
              },
              "type": "relationship",
            } 
        producer.send(kafka_topic, value=data)
    except Exception as err:
        print("Error when creating: " +
          task_id +
          " to " +
          dataset_id + 
          " Error message: " +
          err)


async def main():
    dag = "test"  # Replace with the desired dagId
    await create_dag_node(dag)
# Run the main function within an event loop
loop = asyncio.get_event_loop()
loop.run_until_complete(main())

producer.close()
